### DISCUSSION TITLE
Including static banner at the top of every page – how to make it performant?

### DISCUSSION THREAD

**Discussion by DavidChouinard** [vercel/next.js#19494] [Help] [answered]
Hi,

I'm looking to add a statically-generated banner to all pages (content is queried from our CMS at build time). We could add a custom `app.js` with a `getInitialProps` that queries our CMS, however it's my understanding that this will be called for every page individually — greatly slowing down our build time (we have 40k+ pages).

The content does not vary by page, is there a way to query our CMS _once_ and use that to power the relevant component?

**Comment by oscarpas** [comment]
If you use static pages it should only be queried at build time. If you use server side rendered pages then you could add an API route that proxies and caches requests to your CMS.

**Comment by DavidChouinard** [comment, marked as answer]
@oscarpas Not sure I understand, if we use static pages, won't there be the same issue? (every page will require additional processing, greatly increasing the build time)

Looking to setup things so that there's 1 additional API request at build time (not _n_ API request, where _n_ is the number of pages). Is that possible?

**Reply by oscarpas** [reply]
@DavidChouinard Well yes, I figured that wasn't a problem since it's just at build time and doesn't affect the end-user. But I guess it could be problematic when you have 40k pages, missed that part.

Two workarounds I can think of:

1. Create a script that fetches the CMS content and saves it as a JSON file. In your component you import the CMS content from the file instead of querying for it. Run the script before building.

2. Implement some type of cache layer.

**Reply by DavidChouinard** [reply]
@oscarpas Awesome, very clever suggestion. Working flawlessly with a pre-build script that outputs a JSON file to `public/` and the relevant component reads it. Thank you!
